### PR TITLE
Add chinese support into BERT pretraining

### PR DESCRIPTION
**Pull request by qijiaxing**
Fixed problems in BERT pretraining dataset layer to support Chinese. Added comments to make code easier to understand as well.

To validate these changes, I trained on two datasets: english wikitext and chinese wiki_zh. Both trainings converged as expected. No errors were raised during training.

### CODE REVIEW

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 1]
```diff
@@ -140,71 +140,58 @@ def __getitem__(self, idx, min_doc_length=16):
             # TODO: maybe introduce an argument to control this.
             target_seq_length = random.randint(2, max_num_tokens)
 
-        a_filename = random.choice(self.filenames)
-        a_line_idx = random.choice(
-            range(len(self.sentence_indices[a_filename])))
+        # prefer the seq_a to be slightly longer than seq_b, hence 0.6
+        # TODO: perhaps give this as an argument to control.
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: use an argument with default value = 0.6

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 2]
```diff
@@ -216,81 +203,71 @@ def match_target_seq_length(document, target_seq_length, filename,
...
+        is_next = int(not take_random_b)
+        b_line_pos = self.sentence_indices[b_filename][b_line_idx]
+        b_document = get_document(b_filename, b_line_pos)
+        b_document, b_line_idx = match_target_seq_length(
+            b_document, target_seq_length_b, b_filename, b_line_idx,
+            self.sentence_indices)
 
-                trunc_document = a_document if len(
-                    a_document) > len(b_document) else b_document
-                if len(trunc_document) <= 1:
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: this shouldn't be deleted

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 3]
```diff
@@ -216,81 +203,71 @@ def match_target_seq_length(document, target_seq_length, filename,
...
         input_type_ids = np.zeros(self.max_seq_length, dtype=np.int)
         input_type_ids[len(a_document) + 2:len(output_ids) + 1] = 1
 
         while len(input_ids) < self.max_seq_length:
-            input_ids.append(0)
-            output_ids.append(0)
-            output_mask = np.append(output_mask, [0])
+            input_ids.append(self.pad_id)
+            output_ids.append(self.pad_id)
+            output_mask.append(0)
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: extra = max(0, self.max_seq_length - input_ids)
inputs_ids.extend([self.pad_id] * extra)
output_ids.extend([self.pad_id] * extra)
output_mask.extend([0] * extra)

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 4]
```diff
@@ -301,36 +278,41 @@ def mask_ids(self, ids):
         """
 
         # Whole-word masking by default, as it gives better performance.
-        cand_indexes = []
-        for (i, id) in enumerate(ids):
-
-            if len(cand_indexes) >= 1 and not self.tokenizer.ids_to_tokens(
-                    [id])[0].startswith('\u2581'):
+        cand_indexes = [[ids[0]]]
+        for id in ids[1:]:
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: id is a reserved keyword

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 5]
```diff
@@ -35,6 +35,9 @@ def __init__(self,
                  short_seq_prob=0.1,
                  sentence_idx_file=None):
         self.tokenizer = tokenizer
+        self.cls_id = tokenizer.token_to_id("[CLS]")
+        self.sep_id = tokenizer.token_to_id("[SEP]")
+        self.pad_id = tokenizer.token_to_id("[PAD]")
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: nice

Comment (qijiaxing) [reply collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: thanks!

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py (hunk 1 above)
Comment (qijiaxing) [reply collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: yes, added an argument `seq_a_ratio` with default value 0.6

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py (hunk 2 above)
Comment (qijiaxing) [reply collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: yes. added back the length checking.

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py (hunk 3 above)
Comment (qijiaxing) [reply collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: yes. improved as you suggested.

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py (hunk 4 above)
Comment (qijiaxing) [reply collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: used `tid` instead.

**Comment by qijiaxing** [comment]
Added `seq_a_ratio` as an argument into `BertPretrainingDataset` and also some other improvements as @chiphuyen suggested.

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 6]
```diff
@@ -140,71 +141,57 @@ def __getitem__(self, idx, min_doc_length=16):
...
-        if self.sentence_indices[a_filename][a_line_idx] >= \
-                self.sentence_indices[a_filename][-1] or \
-                random.random() < 0.5:
-
-            # About 50% of the time, B is a random sentence from the corpus
-            label = 0
+        is_last_line = \
+            a_line_idx >= (len(self.sentence_indices[a_filename]) - 1)
+        # About 50% of the time, B is a random sentence from the corpus
+        take_random_b = (random.random() < 0.5) or is_last_line
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: nice

File: collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py [hunk 7]
```diff
@@ -216,81 +203,75 @@ def match_target_seq_length(document, target_seq_length, filename,
...
+        if padding_length > 0:
+            input_ids.extend([self.pad_id] * padding_length)
+            output_ids.extend([self.pad_id] * padding_length)
+            output_mask.extend([0] * padding_length)
 
         # TODO: wrap the return value with () for consistent style.
         return np.array(input_ids), input_type_ids,\
             np.array(input_mask, dtype=np.float32), np.array(output_ids),\
-            np.array(output_mask, dtype=np.float32), label
+            np.array(output_mask, dtype=np.float32), is_next
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py]: a small style change to make it easier to read:
return (np.array(input_ids),
            input_type_ids,
            np.array(input_mask, dtype=np.float32),
            np.array(output_ids),
            np.array(input_mask, dtype=np.float32),
            np.array(output_ids),
            np.array(output_mask, dtype=np.float32),
            label,
            np.array(output_mask, dtype=np.float32),
            is_next)

**Pull request merged**